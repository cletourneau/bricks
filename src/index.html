<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Bricks</title>
    <link rel="stylesheet" href="assets/style/styles.css" />
    <script type="module" src="app.js"></script>
</head>
<body>
<h1>Hello Bricks!</h1>
</body>
</html>