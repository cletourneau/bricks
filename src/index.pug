doctype html
html(lang="en")
    head
        meta(charset="UTF-8")
        title Bricks
        link(rel="stylesheet", href="assets/style/styles.css")
        script(type="module", src="app.js")
    body
        h1 Game of Bricks!
        canvas#gameCanvas(width=1600, height=900)