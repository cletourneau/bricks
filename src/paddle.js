export const BOTTOM_MARGIN = 10

export class Paddle {
    constructor(elementId, canvasSize) {
        this.image = document.getElementById(elementId);
        // this.width = this.image.naturalWidth;
        this.height = this.image.naturalHeight;
        this.centerPosition = this.startingPosition(canvasSize);
    }

    // update(context) {
    //     context.drawImage(
    //         this.image,
    //         this.centerPosition.x - this.width / 2,
    //         this.centerPosition.y - this.height / 2,
    //         this.width,
    //         this.height
    //     );
    // }

    startingPosition(canvasSize) {
        return {x: canvasSize.width / 2, y: canvasSize.height - BOTTOM_MARGIN - (this.height / 2)};
    }

    // move(newXPosition) {
    //     this.centerPosition.x = newXPosition
    // }
}