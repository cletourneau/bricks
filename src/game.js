import {Paddle} from "./paddle";

export class Game {
    constructor(elementIds) {
        const canvas = document.getElementById(elementIds.canvas)
        this.context = canvas.getContext('2d')
        this.canvasSize = {width: canvas.clientWidth, height: canvas.clientHeight}
        this.paddle = new Paddle(elementIds.paddle, this.canvasSize);
        this.listenForMouseMovement(canvas);
    }

    listenForMouseMovement(canvas) {
        canvas.addEventListener('mousemove', e => {
            this.paddle.move(e.offsetX)
        })
    }

    update(delta) {
        this.paddle.update(delta)
    }

    redraw() {
        this.context.clearRect(0, 0, this.canvasSize.width, this.canvasSize.height);
        this.paddle.redraw(this.context)
    }
}