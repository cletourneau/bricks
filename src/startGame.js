import {Game} from "./game";


export function start() {
    let game = new Game({canvas: "gameCanvas", paddle: "paddle"})

    let latestAnimate = performance.now()

    function animate(timestamp) {
        let delta = timestamp - latestAnimate
        game.update(delta)
        game.redraw()

        latestAnimate = timestamp
        requestAnimationFrame(animate)
    }

    requestAnimationFrame(animate)
}