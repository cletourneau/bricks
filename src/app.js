import {start} from "./startGame"

start();