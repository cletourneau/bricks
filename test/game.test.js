import {Game} from "../src/game"
import {Paddle} from "../src/paddle"

jest.mock('../src/paddle')

describe('the game', () => {
    const elementIds = {
        canvas: "gameCanvas",
        paddle: "paddle"
    }

    describe('when created', () => {
        const mockAddEventListener = jest.fn()
        const mockGetElementById = jest.fn(() => {
            return {clientWidth: 400, clientHeight: 200, getContext: jest.fn(), addEventListener: mockAddEventListener}
        })

        beforeAll(() => {
            document.getElementById = mockGetElementById
        })

        it('creates a paddle with element id and canvas size', () => {
            new Game(elementIds)

            expect(mockGetElementById.mock.calls[0][0]).toEqual(elementIds.canvas)
            expect(Paddle).toHaveBeenCalledWith(elementIds.paddle, {width: 400, height: 200})
        })

        it('registers for mouse moves', () => {
            const paddleMove = jest.fn()
            Paddle.mockImplementation(() => {
                return {move: paddleMove}
            })

            new Game(elementIds)

            expect(mockAddEventListener).toHaveBeenCalledWith('mousemove', expect.anything())
            const mouseMove = mockAddEventListener.mock.calls[0][1]

            mouseMove({offsetX: 100})

            expect(paddleMove).toHaveBeenCalledWith(100)

        })
    })

    describe('when updated', () => {
        it('updates the paddle', () => {
            const paddleUpdate = jest.fn()
            Paddle.mockImplementation(() => {
                return {update: paddleUpdate}
            })

            new Game(elementIds).update(33)
            expect(paddleUpdate).toHaveBeenCalledWith(33)
        })
    })

    describe('when redrawed', () => {
        const canvas = document.createElement('canvas')
        const context = canvas.getContext('2d')
        const paddleRedraw = jest.fn()

        beforeAll(() => {
            document.getElementById = jest.fn(() => {
                return canvas
            })

            Paddle.mockImplementation(() => {
                return {redraw: paddleRedraw}
            })
        })

        it('clears the game canvas', () => {
            const game = new Game(elementIds)
            forceCanvasSize(game, {width: 800, height: 600})
            game.redraw()

            const events = context.__getEvents()
            expect(events).toMatchSnapshot()
        });

        it('redraws the paddle with the canvas context', () => {
            new Game(elementIds).redraw()
            expect(paddleRedraw).toHaveBeenCalledWith(context)
        })

        function forceCanvasSize(game, size) {
            game.canvasSize = size
        }
    })
})

