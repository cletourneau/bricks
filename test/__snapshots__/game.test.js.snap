// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the game when redrawed clears the game canvas 1`] = `
Array [
  Object {
    "props": Object {
      "height": 600,
      "width": 800,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
]
`;
